pkg/resource/deploy/step.go

```diff
@@ -1260,24 +1265,37 @@ func (s *ImportStep) Op() display.StepOp {
 	return OpImport
 }
 
-func (s *ImportStep) Deployment() *Deployment                      { return s.deployment }
-func (s *ImportStep) Type() tokens.Type                            { return s.new.Type }
-func (s *ImportStep) Provider() string                             { return s.new.Provider }
-func (s *ImportStep) URN() resource.URN                            { return s.new.URN }
-func (s *ImportStep) Old() *resource.State                         { return s.old }
-func (s *ImportStep) New() *resource.State                         { return s.new }
-func (s *ImportStep) Res() *resource.State                         { return s.new }
-func (s *ImportStep) Logical() bool                                { return !s.replacing }
-func (s *ImportStep) Diffs() []resource.PropertyKey                { return s.diffs }
-func (s *ImportStep) DetailedDiff() map[string]plugin.PropertyDiff { return s.detailedDiff }
+func (s *ImportStep) Deployment() *Deployment { return s.deployment }
+func (s *ImportStep) Type() tokens.Type       { return s.new.Type }
+func (s *ImportStep) Provider() string        { return s.new.Provider }
+func (s *ImportStep) URN() resource.URN       { return s.new.URN }
+func (s *ImportStep) Old() *resource.State    { return s.old }
+func (s *ImportStep) New() *resource.State    { return s.new }
+func (s *ImportStep) Res() *resource.State    { return s.new }
+func (s *ImportStep) Logical() bool           { return !s.replacing }
+
+func (s *ImportStep) Apply() (_ resource.Status, _ StepCompleteFunc, err error) {
+	defer func() {
+		// Ensure that we reject the completion source if we fail to complete the import.
+		if err != nil && s.cts != nil {
+			s.cts.MustReject(err)
+		}
+	}()
 
-func (s *ImportStep) Apply() (resource.Status, StepCompleteFunc, error) {
 	complete := func() {
-		s.reg.Done(&RegisterResult{State: s.new})
+		if s.cts != nil {
+			s.cts.MustFulfill(s.new)
+		}
+		if s.planned {
+			s.reg.Done(&RegisterResult{State: s.new})
+		}
```

Comment:
I feel like this second `if` warrants a comment. I.e. why aren't we `Done` in non-planned?